File: python/2-D Dynamic Programming/burst_balloons.py
from typing import List

class Solution:
    """
    Burst Balloons - Maximum Coins
    
    Problem: Given n balloons, each with a number on it, burst them to maximize coins.
    When you burst balloon i, you get nums[left] * nums[i] * nums[right] coins, where
    left and right are the adjacent balloons (not yet burst). The boundaries are treated
    as 1 (i.e., nums[-1] = nums[n] = 1). Return the maximum coins you can collect.
    
    Example: nums = [3,1,5,8]
    Output: 167
    Explanation: Burst order 1, 5, 3, 8 yields 167 coins.
    
    Approach: Dynamic Programming with Memoization (Interval DP)
    - Add 1 to both ends of nums to handle boundaries
    - Use DFS with memoization to try all possible last balloons to burst in interval [l, r]
    - For each interval, try bursting each balloon last and recursively solve subproblems
    - Memoize results to avoid recomputation
    - State: (l, r) = max coins for bursting all balloons in interval [l, r]
    
    Time Complexity: O(n^3) where n = len(nums)
    Space Complexity: O(n^2) for memoization cache
    """
    
    def maxCoins(self, nums: List[int]) -> int:
        # Add 1 to both ends to handle boundaries
        nums = [1] + nums + [1]
        dp = {}  # Memoization cache: (l, r) -> max coins for interval [l, r]

        def dfs(l, r):
            """
            DFS helper to compute max coins for bursting balloons in interval [l, r]
            Args:
                l: left index (inclusive)
                r: right index (inclusive)
            Returns:
                Maximum coins for bursting all balloons in [l, r]
            """
            # Base case: no balloons to burst
            if l > r:
                return 0
            # Return cached result if already computed
            if (l, r) in dp:
                return dp[(l, r)]

            max_coins = 0
            # Try bursting each balloon in [l, r] last
            for i in range(l, r + 1):
                # Coins from bursting balloon i last in this interval
                coins = nums[l - 1] * nums[i] * nums[r + 1]
                # Add coins from subproblems (left and right intervals)
                coins += dfs(l, i - 1) + dfs(i + 1, r)
                max_coins = max(max_coins, coins)
            dp[(l, r)] = max_coins
            return max_coins

        # Start with the interval covering the original nums
        return dfs(1, len(nums) - 2)


# Test cases
def test_burst_balloons():
    """Test cases for Burst Balloons problem"""
    
    # Test case 1: Basic case
    print("Test 1: nums=[3,1,5,8]")
    sol = Solution()
    result = sol.maxCoins([3,1,5,8])
    expected = 167  # Optimal burst order yields 167 coins
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 1 passed\n")
    
    # Test case 2: Single balloon
    print("Test 2: nums=[1]")
    result = sol.maxCoins([1])
    expected = 1  # Only one balloon, coins = 1*1*1
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 2 passed\n")
    
    # Test case 3: Two balloons
    print("Test 3: nums=[1,5]")
    result = sol.maxCoins([1,5])
    expected = 10  # Burst 1 then 5: 1*1*5 + 1*5*1 = 5 + 5 = 10
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 3 passed\n")
    
    # Test case 4: All ones
    print("Test 4: nums=[1,1,1,1]")
    result = sol.maxCoins([1,1,1,1])
    expected = 4  # Each burst gives 1, total 4
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 4 passed\n")
    
    # Test case 5: Empty array
    print("Test 5: nums=[]")
    result = sol.maxCoins([])
    expected = 0  # No balloons to burst
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 5 passed\n")
    
    # Test case 6: Increasing values
    print("Test 6: nums=[1,2,3,4,5]")
    result = sol.maxCoins([1,2,3,4,5])
    expected = 110  # Optimal burst order yields 110 coins
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 6 passed\n")
    
    # Test case 7: Large values
    print("Test 7: nums=[9,76,64,21,97]")
    result = sol.maxCoins([9,76,64,21,97])
    # We don't check expected value here, just ensure no error and result is int
    print(f"Result: {result}, Type: {type(result)}")
    assert isinstance(result, int)
    print("✓ Test 7 passed\n")
    
    print("All tests passed!")


if __name__ == "__main__":
    test_burst_balloons()

File: C++/2-D Dynamic Programming/burst_balloons.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Function to compute the maximum coins obtainable by bursting balloons optimally.
     *
     * Approach: Interval DP + Memoization
     * - Pad the input array with 1 at both ends to handle boundaries easily.
     * - Use DFS + memoization to compute max coins in interval [l, r].
     * - For each interval, try bursting every balloon last and recursively solve subintervals.
     *
     * Time Complexity: O(n^3) where n = nums.size()
     * Space Complexity: O(n^2) for the DP cache
     */
    int maxCoins(vector<int>& nums) {
        int n = nums.size();

        // Pad nums with 1 at both ends
        nums.insert(nums.begin(), 1);
        nums.push_back(1);

        // Memoization cache: map of pair<int,int> -> int
        unordered_map<long long, int> dp;

        /**
         * Helper function to compute max coins in interval [l, r]
         * @param l - left index (inclusive)
         * @param r - right index (inclusive)
         * @param nums - padded vector of balloons
         * @param dp - memoization cache
         * @return maximum coins obtainable in [l, r]
         */
        function<int(int, int)> dfs = [&](int l, int r) -> int {
            if (l > r) return 0;

            // Encode (l,r) as a single key (since unordered_map doesn't support pair keys directly)
            long long key = ((long long)l << 32) | r;

            if (dp.find(key) != dp.end()) return dp[key];

            int max_coins = 0;

            // Try each balloon in [l,r] as the last balloon to burst
            for (int i = l; i <= r; ++i) {
                int coins = nums[l-1] * nums[i] * nums[r+1];
                coins += dfs(l, i-1) + dfs(i+1, r);
                max_coins = max(max_coins, coins);
            }

            dp[key] = max_coins;
            return max_coins;
        };

        return dfs(1, n);
    }
};

/*
===================================================
Line-by-line C++ syntax explanation:
===================================================

#include <bits/stdc++.h>
    -> Includes all standard C++ libraries (like <vector>, <unordered_map>, <string>, etc.)

using namespace std;
    -> Allows using standard library names (like vector, unordered_map) without prefixing them with std::

class Solution {
public:
    -> Defines a class named Solution with public member functions.

int maxCoins(vector<int>& nums) {
    -> Defines a member function `maxCoins` that takes a vector of ints (passed by reference) and returns an int.

int n = nums.size();
    -> Stores the size of the input vector in `n`.

nums.insert(nums.begin(), 1);
    -> Inserts 1 at the beginning of the vector `nums`.

nums.push_back(1);
    -> Appends 1 to the end of the vector `nums`.

unordered_map<long long, int> dp;
    -> Declares a hash map to store computed results for intervals.
       Uses `long long` as key (to encode pair<int,int>) and `int` as value.

function<int(int, int)> dfs = [&](int l, int r) -> int { ... };
    -> Defines a lambda function named `dfs` with capture [&], which takes (l, r) and returns an int.

if (l > r) return 0;
    -> Base case: if the interval is invalid, no balloons left to burst.

long long key = ((long long)l << 32) | r;
    -> Encodes the pair (l, r) into a single `long long` key.

if (dp.find(key) != dp.end()) return dp[key];
    -> If result for interval [l,r] is already computed, return cached result.

int max_coins = 0;
    -> Initialize variable to track maximum coins in current interval.

for (int i = l; i <= r; ++i) { ... }
    -> Try bursting each balloon `i` last in the interval [l,r].

int coins = nums[l-1] * nums[i] * nums[r+1];
    -> Coins gained by bursting balloon `i` last in this interval.

coins += dfs(l, i-1) + dfs(i+1, r);
    -> Add the coins from left and right subintervals.

max_coins = max(max_coins, coins);
    -> Keep track of maximum coins obtained among all choices.

dp[key] = max_coins;
    -> Memoize the computed result for interval [l,r].

return dp[key];
    -> Return the result.

return dfs(1, n);
    -> Start the recursion for interval [1, n] (excluding the added boundaries).

===================================================
Time Complexity:
    - There are O(n^2) possible intervals [l,r], and each requires O(n) work → O(n^3).

Space Complexity:
    - The DP cache stores O(n^2) entries → O(n^2) space.

===================================================
*/

